cmd/mrthreshold.cpp

```diff
@@ -14,222 +14,442 @@
+    + Argument ("value").type_float()
+
+  + Option ("percentile", "determine threshold based on some percentile of the image intensity distribution")
+    + Argument ("value").type_float (0.0, 100.0)
+
+  + Option ("top", "determine threshold that will result in selection of some number of top-valued voxels")
+    + Argument ("count").type_integer (1)
+
+  + Option ("bottom", "determine & apply threshold resulting in selection of some number of bottom-valued voxels "
+                      "(note: implies threshold application operator of \"le\" unless otherwise specified)")
+    + Argument ("count").type_integer (1)
+
+
 
-  + Option ("top", "provide a mask of the N top-valued voxels")
-  + Argument ("N").type_integer (0)
+  + OptionGroup ("Threshold determination modifiers")
 
-  + Option ("bottom", "provide a mask of the N bottom-valued voxels")
-  + Argument ("N").type_integer (0)
+  + Option ("allvolumes", "compute a single threshold for all image volumes, rather than an individual threshold per volume")
 
-  + Option ("invert", "invert output binary mask.")
+  + Option ("ignorezero", "ignore zero-valued input values during threshold determination")
 
-  + Option ("toppercent", "provide a mask of the N%% top-valued voxels")
-  + Argument ("N").type_float (0.0, 100.0)
+  + Option ("mask", "compute the threshold based only on values within an input mask image")
+    + Argument ("image").type_image_in ()
 
-  + Option ("bottompercent", "provide a mask of the N%% bottom-valued voxels")
-  + Argument ("N").type_float (0.0, 100.0)
 
-  + Option ("nan", "use NaN as the output zero value.")
 
-  + Option ("ignorezero", "ignore zero-valued input voxels.")
+  + OptionGroup ("Threshold application modifiers")
+
+  + Option ("operator", "mathematical operator to use when applying the threshold; "
+                        "options are: " + join(operator_list, ",") + " (default = ge)")
+    + Argument ("choice").type_choice (operator_list)
+
+  + Option ("nan", "set voxels that fail the threshold to NaN rather than zero "
+                   "(output image will be floating-point rather than binary)");
 
-  + Option ("mask", "compute the optimal threshold based on voxels within a mask.")
-  + Argument ("image").type_image_in ();
 }
 
 
-void run ()
-{
-  default_type threshold_value (NaN), percentile (NaN), bottomNpercent (NaN), topNpercent (NaN);
-  size_t topN (0), bottomN (0), nopt (0);
+using value_type = float;
+
+
+
+bool issue_degeneracy_warning = false;
+
 
-  auto opt = get_options ("abs");
-  if (opt.size()) {
-    threshold_value = opt[0][0];
-    ++nopt;
-  }
 
-  opt = get_options ("percentile");
+Image<bool> get_mask (const Image<value_type>& in)
+{
+  Image<bool> mask;
+  auto opt = get_options ("mask");
   if (opt.size()) {
-    percentile = opt[0][0];
-    ++nopt;
+    mask = Image<bool>::open (opt[0][0]);
+    if (mask.ndim() > in.ndim())
+      throw Exception ("Cannot use mask image with more axes than input image");
```

Comment:
This won't allow e.g. using a 4D mask with a single volume (effectively 3D) with a genuine 3D image. Is this intentional? Maybe we can relax this a bit?

Reply:
05c736e permits that case.